src/lambda/trans.lisp

```diff
@@ -47,6 +47,189 @@ calling me"
+                                          so1
+                                          (maybe-rest (coprod mcartoon
+                                                              mcadrtoon)))
+                              (geb:pair ctx (rec context on))))))
+               ((pair ltm rtm)
+                (let ((lty  (ttype ltm))
+                      (rty  (ttype rtm)))
+                  (comp (->right so1 (prod (maybe-comp lty)
+                                           (maybe-comp rty)))
+                        (geb:pair (rec context ltm)
+                                  (rec context rtm)))))
+               ((fst term)
+                (let ((mcarttot (mcar (ttype term))))
+                  (comp (mcase (->left so1 (maybe-rest mcarttot))
+                               (<-left (maybe-comp mcarttot)
+                                       (maybe (mcadr (ttype term)))))
+                        (rec context term))))
+               ((snd term)
+                (let ((mcadrttot (mcadr (ttype term))))
+                  (comp (mcase (->left so1 (maybe-rest mcadrttot))
+                               (<-right (maybe-comp (mcar (ttype term)))
+                                        (maybe-comp mcadrttot)))
+                        (rec context term))))
+               ((lamb tdom term)
+                (comp (->right so1 (maybe-rest (fun-to-hom (ttype tterm))))
+                      (apply-n (length tdom)
+                               #'(lambda (x) (curry (commutes-left x)))
+                               (rec (append
+                                     (mapcar #'maybe-comp tdom)
+                                     context) term))))
+               ((app fun term)
+                (let ((tofun (ttype fun)))
+                  (comp
+                   (mcase (->left so1 (maybe-rest tofun))
+                          (commutes-left
+                           (so-eval (maybe-comp (mcar tofun))
+                                    (maybe-comp (mcadr tofun)))))
+                   (comp (distribute (reduce #'prod
+                                             (mapcar #'maybe-comp term)
+                                             :from-end t)
+                                     so1
+                                     (maybe-rest tofun))
+                         (geb:pair (reduce #'geb:pair
+                                           (mapcar #'(lambda (x) (rec context x)) term)
+                                           :from-end t)
+                                   (rec context
+                                        fun))))))
+               ((index pos)
+                (stlc-ctx-proj (mapcar #'maybe context) pos))
+               ((err ttype)
+                (comp (->left so1 (maybe-comp ttype))
+                      (terminal (stlc-ctx-maybe context)))))))
+    (if (not (well-defp context tterm))
+        (error "not a well-defined ~A in said ~A" tterm context)
+        (rec context (ann-term1 context tterm)))))
+
```

Comment:
The most painful kind of copy and paste.

Let us remove this with a better strategy later, or find a way to abstract it if it has to live here long term